### ISSUE TITLE
Group and Role Use Cases

### ISSUE THREAD

**Issue by vantuyls** [projecthydra-labs/hydra-groupy#1]
Stick 'em here!

***

The overall goal of the design is to develop a UI-based method of assigning a user the ability to perform functions within the application.

However, we recognize there are a broad range of needs, and are working to provide a solution to the above which:
* is adapter-based, with a validated interface, to allow for personalized solutions
* consolidates the abilities to one location within the application for clarity, while atomizing the definition of those abilities (DRY)
* clarifies the difference between groups and roles, and allows for limited implementations for those who don't need all of the options
* maintains backward compatibility
* is well documented to allow for for community understanding and input

By including your use cases below, our goal is to account for them in our solution, either through implementing them as a feature, or by providing documented points for you to personalize your own implementation.

**Comment by tpendragon** [comment]
We want to spin up our application and have permissions "just work" - basically, all the logic sits in our code somewhere. We don't want to have to assign permissions from a database for every user + action.

**Comment by vantuyls** [comment]
1) As a repository administrator, i want to be able to create a group of users and allow them to only deposit to a specific work type.

2) As a repository administrator, i only want one place in the interface where I am managing groups and what the members of that group are allowed to do (as opposed to currently in Hyrax, where i have to manage system-wide groups/roles and groups/roles on a workflow and on an admin set).

**Comment by revgum** [comment]
It’d be nice if in the UI when a users set of functional abilities is displayed you can see what is applying the ability and what the “final” set of abilities would be.. so for instance, when there are overlapping abilities assigned you can get some awareness of which thing had the “final say” .. so to speak

**Comment by elrayle** [comment]
**Groups:**  one per department (e.g. Philosophy, Math, English, etc.)
**Functional Roles:** super_admin, admin, depositor, reviewer, viewer
**Collectors:** create a Collection (or AdminSet) per department giving it the name of the department (e.g. Philosophy Department, Math Department, English Department, etc.)
**Abilities:** 
* As super_admin, I can do everything to everything.
* As admin for the Math group, I can...
  * manage: Math Department Collection
  * deposit in: Math Department Collection
  * am granted edit_access to all works in the Math Department Collection
  * etc.
* As depositor for the Math group, I can...
  * deposit in: Math Department Collection
  * manage: works I deposit
etc.

The key point is that as a depositor for the Math group, I cannot deposit to the Philosophy Department collection, English Department collection or any other collection.

Ideally, when a collection is created, it can be associated with a group and have sets of roles with abilities defined for the roles that apply to that collection only.  The same roles and defined abilities can be used with a different collection based on membership in its associated group.

**Comment by laritakr** [comment]
From Maria Whitaker (IU): 

What use case would fit the scenario of default behavior with respect to changing the Admin Set manager from one person to another and not changing the manager from the items in the set. The use case I had in mind would not work in this scenario: I am the manager of Admin Set 1 and I quit my job. My employer changes the manager of Admin Set 1 to be my successor. My successor has no access to the items in the Admin Set unless we programmatically go and change the manager on those items.

After some more thought, I concluded that my case could be handled by creating a group named “Manager of Admin Set 1”, adding me to that group, and setting up that group to be the Manager of the Admin Set 1. This way, when I quit, my successor can replace me in the group and thus my successor has management access to all items I managed before.

**Comment by elrayle** [comment]
RE: Maria Whitaker - clarification

Unless something changes, a Manager of an AdminSet is given edit_access to all works in the AdminSet **at the time the work is created in the AdminSet**.  Changing the manager does NOT change the edit_access of existing works in an AdminSet.  Moving a work from one AdminSet to another does NOT change the edit_access of that work.  Any new works created in the AdminSet will give edit_access to the new manager.

Using a  group as the manager of an AdminSet does help ameliorate this as the group is provided edit_access.  If group membership changes, the access controls will be properly applied for the new set of users in the group without requiring change to the works edit_access settings.

**Comment by laritakr** [comment]
@elrayle Yes I would agree that your clarification is correct. I added Maria's comment to have it documented, but I believe that implementing groups and roles appropriately would handle her concerns.

**Comment by laritakr** [comment]
Notre Dame use cases include: 

- ability to grant or change user permissions without requiring a redeploy
- ability to troubleshoot missing access (by somehow comparing what permissions the user has, with the permissions that are needed for the given task)

**Comment by cjcolvar** [comment]
Avalon is moving to Hyrax and would be interested in this work.
Avalon's current roles are pretty well defined in this blog post:
http://www.avalonmediasystem.org/blog-post/understanding-avalon-roles-and-permissions
